### PR TITLE
Declarative bulk subscribe support

### PR DESCRIPTION
**Pull request by DeepanshuA**
Signed-off-by: Deepanshu Agarwal <deepanshu.agarwal1984@gmail.com>

# Description

This piece adds Declarative Support to Bulk Subscribe. It also changes programmatic way, by bringing bulkSubscribe outside metadata, like this(as per consensus reached in #5499 ):
```
apiVersion: dapr.io/v1alpha1
kind: Subscription
metadata:
  name: orders-subscription-mqtt
spec:
  topic: orders
  route: /orders
  pubsubname: mqtt-pubsub
  bulkSubscribe: (optional)
    enabled: true (required, if bulkSubscribe block given)
    maxMessagesCount: 4 (optional)
    maxAwaitDurationMs: 10000 (optional)
  metadata:
    rawPayload: "true"
```
Please note that I haven't put any defaults for maxMessagesCount and maxAwaitDurationMs in runtime, as wanted to leave these defaults on the component side. If we intend to introduce  common defaults across components - even for that, defaults should ideally go in contrib repo.

## Issue reference

Please reference the issue this PR will close: #5499 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: dapr/proto/runtime/v1/appcallback.proto [hunk 1]
```diff
@@ -287,6 +290,18 @@ message TopicRule {
   string path = 2;
 }
 
+// BulkSubscribeRequest is the message to pass settings for bulk subscribe
+message BulkSubscribeRequest {
+  // Required. Flag to enable/disable bulk subscribe
+  string enabled = 1;
+
+  // Optional. Max number of messages to be sent in a single bulk request
+  string maxBulkSubCount = 2;
```
Comment (artursouza) [inline dapr/proto/runtime/v1/appcallback.proto:299]: maxMessagesCount since this property is already in BulkSubscribeRequest, having "BulkSub" is redundant.

File: dapr/proto/runtime/v1/appcallback.proto [hunk 2]
```diff
@@ -287,6 +290,18 @@ message TopicRule {
...
+
+  // Optional. Max duration to wait for messages to be sent in a single bulk request
+  string maxBulkSubAwaitDurationMs  = 3;
```
Comment (artursouza) [inline dapr/proto/runtime/v1/appcallback.proto:302]: Same: maxAwaitDurationMs

File: pkg/apis/subscriptions/v1alpha1/types.go [hunk 3]
```diff
@@ -38,9 +38,18 @@ type SubscriptionSpec struct {
...
 	Route           string            `json:"route"`
+	BulkSubscribe   BulkSubscribe     `json:"bulkSubscribe,omitempty"`
 	DeadLetterTopic string            `json:"deadLetterTopic,omitempty"`
 }
 
+// BulkSubscribe encapsulates the bulk subscription configuration for a topic.
+type BulkSubscribe struct {
+	Enabled string `json:"enabled"`
+	// +optional
+	MaxBulkSubCount           string `json:"maxBulkSubCount,omitempty"`
```
Comment (artursouza) [inline pkg/apis/subscriptions/v1alpha1/types.go:49]: Same.

File: pkg/apis/subscriptions/v1alpha1/types.go [hunk 4]
```diff
@@ -38,9 +38,18 @@ type SubscriptionSpec struct {
...
+	MaxBulkSubAwaitDurationMs string `json:"maxBulkSubAwaitDurationMs,omitempty"`
```
Comment (artursouza) [inline pkg/apis/subscriptions/v1alpha1/types.go:50]: Same.

**Review by artursouza — COMMENTED** [review]
Overall LGTM. Just minor comments to naming.

File: dapr/proto/runtime/v1/appcallback.proto (hunk 1 above)
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto]: Done.

File: pkg/apis/subscriptions/v1alpha1/types.go (hunk 3 above)
Comment (DeepanshuA) [reply pkg/apis/subscriptions/v1alpha1/types.go]: Done.

**Comment by DeepanshuA** [comment]
> Overall LGTM. Just minor comments to naming.

I just realized that I had forgot to raise corresponding PR in components-contrib to take MaxMessagesCount and MaxAwaitDurationMs. I have raised https://github.com/dapr/components-contrib/pull/2405 to take care of that, and this PR can be ONLY AFTER THAT PR IS MERGED. Also, I will have to remove replace from go.mod here.

File: pkg/apis/subscriptions/v1alpha1/types.go (hunk 4 above)
Comment (DeepanshuA) [reply pkg/apis/subscriptions/v1alpha1/types.go]: Done.

File: go.mod [hunk 5]
```diff
@@ -422,7 +422,7 @@ replace (
 // Uncomment for local development for testing with changes in the components-contrib repository.
 // Don't commit with this uncommented!
 //
-// replace github.com/dapr/components-contrib => ../components-contrib
-//
+replace github.com/dapr/components-contrib => github.com/DeepanshuA/components-contrib v1.6.0-rc.2.0.20230104154635-e891f15b6403
```
Comment (DeepanshuA) [inline go.mod:425]: Reviewers: I will be removing this after https://github.com/dapr/components-contrib/pull/2405 is merged. And, this PR has to be merged only after it is removed from here.

File: dapr/proto/runtime/v1/appcallback.proto [hunk 6]
```diff
@@ -287,6 +290,18 @@ message TopicRule {
   string path = 2;
 }
 
+// BulkSubscribeRequest is the message to pass settings for bulk subscribe
+message BulkSubscribeRequest {
+  // Required. Flag to enable/disable bulk subscribe
+  string enabled = 1;
+
+  // Optional. Max number of messages to be sent in a single bulk request
+  string maxMessagesCount = 2;
```
Comment (artursouza) [inline dapr/proto/runtime/v1/appcallback.proto:299]: The convention used in this file is not camelCase but snake_case. See other properties.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: charts/dapr/crds/subscription.yaml [hunk 7]
```diff
@@ -51,6 +51,18 @@ spec:
                 type: string
               deadLetterTopic:
                 type: string
+              bulkSubscribe:
+                description: Represents bulk subscribe properies
+                properties:
+                  enabled: 
+                    type: string
```
Comment (ItalyPaleAle) [inline charts/dapr/crds/subscription.yaml:58]: Should this be a boolean?

File: charts/dapr/crds/subscription.yaml [hunk 8]
```diff
@@ -51,6 +51,18 @@ spec:
...
+                  maxMessagesCount:
+                    type: string
```
Comment (ItalyPaleAle) [inline charts/dapr/crds/subscription.yaml:60]: Should this be an int?

File: charts/dapr/crds/subscription.yaml [hunk 9]
```diff
@@ -51,6 +51,18 @@ spec:
...
+                  maxAwaitDurationMs:
+                    type: string
```
Comment (ItalyPaleAle) [inline charts/dapr/crds/subscription.yaml:62]: Should this be an int?

File: charts/dapr/crds/subscription.yaml [hunk 10]
```diff
@@ -129,6 +141,18 @@ spec:
               deadLetterTopic:
                 description: The optional dead letter queue for this topic to send events to.
                 type: string
+              bulkSubscribe:
+                description: Represents bulk subscribe properies
+                properties:
+                  enabled: 
+                    type: string
```
Comment (ItalyPaleAle) [inline charts/dapr/crds/subscription.yaml:148]: Here too, I think the types should be more explicit: bool, int, int?

File: charts/dapr/crds/subscription.yaml [hunk 11]
```diff
@@ -51,6 +51,18 @@ spec:
...
+                  maxMessagesCount:
+                    type: string
+                  maxAwaitDurationMs:
+                    type: string
+                required:
+                  - enabled
```
Comment (ItalyPaleAle) [inline charts/dapr/crds/subscription.yaml:64]: Why is this required? Can we assume that it's `false` if unspecified?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: charts/dapr/crds/subscription.yaml (hunk 11 above)
Comment (DeepanshuA) [reply charts/dapr/crds/subscription.yaml:64]: Basically, `bulkSubscribe` block in itself is optional. So, if it is not required, it shouldn't / wouldn't be given at all.
But, if someone provide `bulkSubscribe` block, then inside it, `enabled` is mandatory.
Does that sound fine?

File: dapr/proto/runtime/v1/appcallback.proto (hunk 6 above)
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto]: My bad, I have updated.

File: charts/dapr/crds/subscription.yaml (hunk 7 above)
Comment (DeepanshuA) [reply charts/dapr/crds/subscription.yaml]: I had considered everything string and then converted when actually required, but, it makes it better to consider the actual data types. Changed.

File: charts/dapr/crds/subscription.yaml (hunk 8 above)
Comment (DeepanshuA) [reply charts/dapr/crds/subscription.yaml]: Updated

File: charts/dapr/crds/subscription.yaml (hunk 9 above)
Comment (DeepanshuA) [reply charts/dapr/crds/subscription.yaml]: Updated

File: charts/dapr/crds/subscription.yaml (hunk 10 above)
Comment (DeepanshuA) [reply charts/dapr/crds/subscription.yaml]: Updated

**Comment by DeepanshuA** [comment]
@ItalyPaleAle @artursouza Please re-review. Whenever I request re-review from someone, it automatically removes re-review from others. So, putting a comment.

File: dapr/proto/runtime/v1/appcallback.proto [hunk 12]
```diff
@@ -265,6 +265,9 @@ message TopicSubscription {
 
   // The optional dead letter queue for this topic to send events to.
   string dead_letter_topic = 6;
+
+  // The optional bulk subscribe settings for this topic.
+  BulkSubscribeRequest bulk_subscribe = 7;
```
Comment (mukundansundar) [inline dapr/proto/runtime/v1/appcallback.proto:270]:
```suggestion
  BulkSubscribeConfig bulk_subscribe = 7;
```

Comment (DeepanshuA) [reply dapr/proto/runtime/v1/appcallback.proto]: Changed.

File: go.mod (hunk 5 above)
Comment (DeepanshuA) [reply go.mod]: Removed local fork reference, as https://github.com/dapr/components-contrib/pull/2405 is merged.

**Comment by DeepanshuA** [comment]
> > Overall LGTM. Just minor comments to naming.
> 
> I just realized that I had forgot to raise corresponding PR in components-contrib to take MaxMessagesCount and MaxAwaitDurationMs till component. I have raised [dapr/components-contrib#2405](https://github.com/dapr/components-contrib/pull/2405) to take care of that, and this PR can be merged ONLY AFTER THAT PR IS MERGED. Also, I will have to remove replace from go.mod here.

https://github.com/dapr/components-contrib/pull/2405 is now merged. So, this PR can be in now, once it is Approved.

**Review by ItalyPaleAle — APPROVED** [review]
